keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -0,0 +1,316 @@
+
+Polling incurs non-negligible overhead as the number of pods/containers increases, and is exacerbated by Kubelet's parallelism -- one worker (goroutine) per pod, which queries the container runtime individually. Periodic, concurrent, large number of requests causes high CPU usage spikes (even when there is no spec/state change), poor performance, and reliability problems due to overwhelmed container runtime. Ultimately, it limits Kubelet's scalability.
+
+### Goals
+
+- Reduce unnecessary work during inactivty (no spec/state changes)
+	- In other words, Reduce steady-state CPU usage of Kubelet and CRI implementation by reducing frequent polling.
+
+### Non-Goals
+
+- Completely elimanitate polling altogether.
+    - This proposal does not advocate completely removing the polling. We cannot solely rely on the upstream container events due to the possibility of missing events. PLEG should relist reduced frequency to ensure no events are missed.
+
+## Proposal
+
+This proposal aims to replace the periodic polling with a pod lifecycle/image event watcher. Currently, the Kubelet calls into three CRI calls of the form `List*`: [ListContainers](https://github.com/kubernetes/kubernetes/blob/6efd6582df2011f1ec8c146ef711b3348ae07d60/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L78), [ListPodSandbox](https://github.com/kubernetes/kubernetes/blob/6efd6582df2011f1ec8c146ef711b3348ae07d60/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L60) and [ListImages](https://github.com/kubernetes/kubernetes/blob/6efd6582df2011f1ec8c146ef711b3348ae07d60/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L122). Each of these is used to populate the Kubelet's perspective
+of the state of the node.
+
+As the number of pods on a node increases, the amount of time the Kubelet and CRI implementation takes in generating and reading this list increases linearly. What is needed is a way of the Kubelet being notified when a container or image changes state in a way it did not trigger.
+
+Luckily, there should only be two such cases, and in normal operation, only one would happen frequently:
+
+- The first, and most clear case of a container or image changing state without the Kubelet triggering that state change is when a container stops. Containers can exit gracefully, or be OOM killed, and the Kubelet would not know.
+   - We will also introduce events when the container is created as well as is started. This will help us reduce the relisting that takes placed while the kubelet waits for the container to start.
+   - Although kubelet initiates the container deletion, for sake of increased validation we are also introducing the event to denote that from the runtime.
```

Comment:
and it is possible that the container is deleted by someone calling cri-api directly, in which case kubelet will not be aware.